internal/servers/controller/controller.go

```diff
@@ -123,6 +123,32 @@ func New(ctx context.Context, conf *Config) (*Controller, error) {
 		}
 	}
 
+	clusterListeners := make([]*base.ServerListener, 0)
+	for i := range conf.Listeners {
+		l := conf.Listeners[i]
+		if l == nil || l.Config == nil || l.Config.Purpose == nil {
+			continue
+		}
+		if len(l.Config.Purpose) != 1 {
+			return nil, fmt.Errorf("found listener with multiple purposes %q", strings.Join(l.Config.Purpose, ","))
+		}
+		switch l.Config.Purpose[0] {
+		case "api":
+			c.apiListeners = append(c.apiListeners, l)
+		case "cluster":
+			clusterListeners = append(clusterListeners, l)
+		}
+	}
+	if len(c.apiListeners) == 0 {
+		return nil, fmt.Errorf("no api listeners found")
+	}
+	if len(clusterListeners) != 1 {
```

Comment:
I think it's simpler if you just assign to `c.clusterListener` the first time you encounter a cluster listener and then you just do a nil check, rather than keeping a temporary slice and then assigning out of it.

Reply:
I might be misunderstanding but wouldn't that mean we'd just silently pick one cluster listener from the config? I think we should be explicit here and error if there's more than one.

Or is your suggestion that we use `c.clusterListener` and then nil check inside the loop like this:
```
for i := range conf.Listeners {
		if l == nil || l.Config == nil || l.Config.Purpose == nil {
			continue
		}
		if len(l.Config.Purpose) != 1 {
			return nil, fmt.Errorf("found listener with multiple purposes %q", strings.Join(l.Config.Purpose, ","))
		}
		switch l.Config.Purpose[0] {
		case "api":
			c.apiListeners = append(c.apiListeners, l)
		case "cluster":
                        if c.clusterListener != nil {
                            return nil, fmt.Errorf("more than one cluster listener")
                        }
			c.clusterListener = l
		}
	}
```

Reply:
If you go with the checks in the for loop like that, you need an additional check after the for loop to check if `c.clusterListener` is `nil` and return an error if it is.

Reply:
Yes, that was my suggestion, but Tim is correct. Might as well leave as-is.